### PR TITLE
MXFP8 training bug fixes for quantized_model_init and Torch FSDP fp8 all gather

### PR DESCRIPTION
**Pull request by sudhu2k**
# Description

Ensure `keep_fp8_weight_transpose_cache` flag is set to `True` not only for autocast but also for quantized_model_init.
Fix padding during fp8 all-gather

Fixes: [#15425](https://github.com/ROCm/frameworks-internal/issues/15425)
[#15420](https://github.com/ROCm/frameworks-internal/issues/15420)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py [hunk 1]
```diff
@@ -471,7 +471,10 @@ def __torch_dispatch__(cls, func, types, args, kwargs=None):
             scale_invs = [tensor._rowwise_scale_inv, tensor._columnwise_scale_inv]
             split_sizes_for_scale = [split_size, split_size // MXFP8_BLOCK_SCALING_SIZE]
             # Padding requirements: rowwise dim0 should be divisble by 128, columnwise dim0 should be divisble by 4
-            padding_multiples = [128, 4]
+            # NOTE: ROCm/HIP backend uses an unpadded scale-inv layout (see `MXFP8Quantizer.make_empty`),
+            # so applying the padding here would produce a per-shard scale-inv whose dim-0
+            # does not match the destination scale-inv allocated for the FSDP2 local shard.
+            padding_multiples = [128, 4] if not IS_HIP_EXTENSION else [1, 1]
```
Comment (alextmagro) [inline transformer_engine/pytorch/tensor/mxfp8_tensor.py:520]: I think for gfx1250 we have some other padding requirements, this should be unified with https://github.com/ROCm/TransformerEngine/pull/568

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py [hunk 2]
```diff
@@ -487,7 +490,7 @@ def __torch_dispatch__(cls, func, types, args, kwargs=None):
                 )
                 scale_inv_out = list(scale_inv_out) if scale_inv_out is not None else None
                 # Pad scale_inv_out to be a multiple of pad_multiple
-                if scale_inv_out is not None:
+                if scale_inv_out is not None and pad_multiple > 1:
```
Comment (alextmagro) [inline transformer_engine/pytorch/tensor/mxfp8_tensor.py]: Do we need this additional check? If padding is just 1 (for gfx950<), won't the code have no effect?

**Review by alextmagro — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 2 above)
Comment (sudhu2k) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py]: Yup, you're right. It'll be a no-op. Modified it.

File: transformer_engine/pytorch/tensor/mxfp8_tensor.py (hunk 1 above)
Comment (sudhu2k) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:520]: Agreed. These changes should also be present in that PR accordingly. But I think for now, let's fix the issue on existing archs and make the appropriate changes along with the #568 PR.

Comment (alextmagro) [reply transformer_engine/pytorch/tensor/mxfp8_tensor.py:520]: OK, @matthiasdiener can you work with Sudharshan make sure this is in your PR one way or another?

**Comment by alextmagro** [comment]
LGTM! Just sync with Matthias on that one padding thing please.

**Review by alextmagro — APPROVED** [review]

**Review by Micky774 — APPROVED** [review]

**Review by ipanfilo — APPROVED** [review]
